src/vbk/interpreter.hpp

```diff
@@ -0,0 +1,217 @@
+            }
+            case OP_CHECKATV: {
+                // we do not allow single pop tx simultaneously contain publications and context
+                isPublicationsTx = true;
+                if (isContextTx) {
+                    return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+                }
+
+                // tx has one ATV
+                if (!pub.atv.empty()) {
+                    return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                }
+
+                // validate ATV size
+                const auto& el = stacktop(-1);
+                if (el.size() > config.max_atv_size || el.size() < config.min_atv_size) {
+                    return set_error(serror, SCRIPT_ERR_PUSH_SIZE);
+                }
+
+                // validate ATV content
+                if (with_checks && !pop.checkATVinternally(el)) {
+                    return set_error(serror, SCRIPT_ERR_VBK_ATVFAIL);
+                }
+
+                pub.atv = el;
+                popstack(stack);
+                break;
+            }
+            case OP_CHECKVTB: {
+                // we do not allow single pop tx simultaneously contain publications and context
+                isPublicationsTx = true;
+                if (isContextTx) {
+                    return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+                }
+
+                // validate VTB size
+                const auto& el = stacktop(-1);
+                if (el.size() > config.max_vtb_size || el.size() < config.min_vtb_size) {
+                    return set_error(serror, SCRIPT_ERR_PUSH_SIZE);
+                }
+
+                // validate VTB content
+                if (with_checks && !pop.checkVTBinternally(el)) {
+                    return set_error(serror, SCRIPT_ERR_VBK_VTBFAIL);
+                }
+
+                // tx has many VTBs
+                pub.vtbs.push_back(el);
+                popstack(stack);
+                break;
+            }
+            case OP_CHECKPOP: {
+                // OP_CHECKPOP should be last opcode
+                if (script.GetOp(pc, opcode, vchPushValue)) {
+                    // we could read next opcode. extra opcodes is an error
+                    return set_error(serror, SCRIPT_ERR_VBK_EXTRA_OPCODE);
+                }
+
+                // stack should be empty at this point
+                if (!stack.empty()) {
+                    return set_error(serror, SCRIPT_ERR_EVAL_FALSE);
+                }
+
+                // we did all checks, put true on stack to signal successful execution
+                stack.push_back(vchTrue);
+                break;
+            }
+            default:
+                // forbid any other opcodes in pop transactions
+                return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+            }
+        }
+    } catch (...) {
+        return set_error(serror, SCRIPT_ERR_UNKNOWN_ERROR);
+    }
+
+    if (isPublicationsTx && pub.atv.empty()) {
```

Comment:
Maybe in this line we should provide check that we have only one ATV and not at least one. Because according to the rules of the VeriBlock we should have one ATV in the PopTx

Reply:
ATV can be only one